libs/remote_storage/src/azure_blob.rs

```diff
@@ -335,3 +341,135 @@ impl RemoteStorage for AzureBlobStorage {
+{
+    fn new(inner: S, len: usize) -> NonSeekableStream<S> {
+        use tokio_util::compat::TokioAsyncReadCompatExt;
+
+        let inner = tokio_util::io::StreamReader::new(inner).compat();
+        let inner = Some(inner);
+        let inner = std::sync::Mutex::new(inner);
+        NonSeekableStream::Initial { inner, len }
+    }
+}
+
+impl<S> std::fmt::Debug for NonSeekableStream<S> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Initial { len, .. } => f.debug_struct("Initial").field("len", len).finish(),
+            Self::Actual { len, .. } => f.debug_struct("Actual").field("len", len).finish(),
+            Self::Cloned { len_was, .. } => f.debug_struct("Cloned").field("len", len_was).finish(),
+        }
+    }
+}
+
+impl<S> futures::io::AsyncRead for NonSeekableStream<S>
+where
+    S: futures::stream::Stream<Item = std::io::Result<bytes::Bytes>>,
+{
+    fn poll_read(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+        buf: &mut [u8],
+    ) -> std::task::Poll<std::io::Result<usize>> {
+        match self.project() {
+            NonSeekableStreamProj::Actual {
+                inner, read_any, ..
+            } => {
+                *read_any = true;
+                inner.poll_read(cx, buf)
+            }
+            _ => std::task::Poll::Ready(Err(std::io::Error::new(
+                std::io::ErrorKind::Other,
+                "cloned or initial values cannot be read",
+            ))),
+        }
+    }
+}
+
+impl<S> Clone for NonSeekableStream<S> {
+    fn clone(&self) -> Self {
+        match self {
+            Self::Initial { inner, len } => {
+                if let Some(inner) = inner.lock().unwrap().take() {
+                    Self::Actual {
+                        inner,
+                        len: *len,
+                        read_any: false,
+                    }
+                } else {
+                    Self::Cloned {
+                        inner_was: std::marker::PhantomData,
+                        len_was: *len,
+                    }
+                }
+            }
+            Self::Actual { len, .. } => Self::Cloned {
+                inner_was: std::marker::PhantomData,
+                len_was: *len,
+            },
+            Self::Cloned { inner_was, len_was } => Self::Cloned {
+                inner_was: *inner_was,
+                len_was: *len_was,
+            },
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl<S> azure_core::SeekableStream for NonSeekableStream<S>
+where
+    S: futures::stream::Stream<Item = std::io::Result<bytes::Bytes>>
+        + Unpin
+        + Send
+        + Sync
+        + 'static,
+{
+    async fn reset(&mut self) -> azure_core::error::Result<()> {
+        match self {
+            NonSeekableStream::Initial { .. } => {
+                return Ok(());
+            }
```

Comment:
This is a forward-compatiblity with azure hazard: need to only allow it if the mutex still has something.